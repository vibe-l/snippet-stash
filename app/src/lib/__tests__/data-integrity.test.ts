import { describe, it, expect } from 'vitest';
import { createHash } from 'crypto';
import fs from 'fs';
import path from 'path';
import { execSync } from 'child_process';

const FIXTURES_DIR = path.join(__dirname, 'fixtures');
const SCRIPT_PATH = path.resolve(__dirname, '../../../../scripts/createDocumentIDs.js');

describe('Data file integrity', () => {
  it('should maintain consistent prompts.json structure', () => {
    const filePath = path.join(FIXTURES_DIR, 'prompts.json');
    const content = fs.readFileSync(filePath, 'utf8');
    const hash = createHash('sha256').update(content).digest('hex');
    expect(hash).toMatchSnapshot();
  });

  it('should generate consistent prompts_IDs.json and CSV files', () => {
    const inputPath = path.join(FIXTURES_DIR, 'prompts.json');
    const outputPath = path.join(FIXTURES_DIR, 'prompts_IDs.json');
    const csvPath = path.join(FIXTURES_DIR, 'prompts_doc_word_count.csv');
    
    // Clean up existing CSV file to ensure fresh generation
    if (fs.existsSync(csvPath)) fs.unlinkSync(csvPath);
    
    // Run the script to regenerate files
    execSync(`tsx "${SCRIPT_PATH}" "${inputPath}"`, { stdio: 'inherit' });
    
    // Verify the generated IDs file structure
    const idsContent = fs.readFileSync(outputPath, 'utf8');
    const idsHash = createHash('sha256').update(idsContent).digest('hex');
    expect(idsHash).toMatchSnapshot();
    
    // Verify the generated CSV structure
    const csvContent = fs.readFileSync(csvPath, 'utf8');
    const csvLines = csvContent.split('\n').slice(0, 10); // First 10 lines
    expect(csvLines).toMatchSnapshot();
  });
});

describe('Data structure validation', () => {
  it('should maintain prompts.json schema', () => {
    const filePath = path.join(FIXTURES_DIR, 'prompts.json');
    const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    
    const schema = {
      type: Array.isArray(data) ? 'array' : 'object',
      length: Array.isArray(data) ? data.length : Object.keys(data).length,
      sampleKeys: Array.isArray(data) && data.length > 0 
        ? Object.keys(data[0] || {}).sort() 
        : Object.keys(data).slice(0, 5).sort(),
      firstItemType: Array.isArray(data) && data.length > 0 
        ? typeof data[0] 
        : 'N/A'
    };
    expect(schema).toMatchSnapshot();
  });

  it('should maintain prompts_IDs.json schema', () => {
    const filePath = path.join(FIXTURES_DIR, 'prompts_IDs.json');
    const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    
    const schema = {
      type: Array.isArray(data) ? 'array' : 'object',
      length: Array.isArray(data) ? data.length : Object.keys(data).length,
      sampleKeys: Array.isArray(data) && data.length > 0 
        ? Object.keys(data[0] || {}).sort() 
        : Object.keys(data).slice(0, 5).sort(),
      firstItemType: Array.isArray(data) && data.length > 0 
        ? typeof data[0] 
        : 'N/A'
    };
    expect(schema).toMatchSnapshot();
  });

  it('should maintain CSV header structure', () => {
    const filePath = path.join(FIXTURES_DIR, 'prompts_doc_word_count.csv');
    const content = fs.readFileSync(filePath, 'utf8');
    const lines = content.split('\n');
    const header = lines[0];
    const sampleRowCount = Math.min(5, lines.length - 1);
    
    const csvStructure = {
      header: header.split(','),
      totalRows: lines.length - 1, // Excluding header
      sampleRowCount,
      hasValidRows: lines.slice(1, sampleRowCount + 1).every(line => 
        line.split(',').length === header.split(',').length
      )
    };
    expect(csvStructure).toMatchSnapshot();
  });

  it('should track data file counts and sizes', () => {
    const files = [
      'prompts.json',
      'prompts_IDs.json', 
      'prompts_doc_word_count.csv'
    ];
    
    const stats = files.map(filename => {
      const filePath = path.join(FIXTURES_DIR, filename);
      const stats = fs.statSync(filePath);
      return {
        filename,
        sizeBytes: stats.size,
        sizeCategory: stats.size < 1024 ? 'small' 
          : stats.size < 1024 * 1024 ? 'medium' : 'large'
      };
    });
    
    expect(stats).toMatchSnapshot();
  });
});