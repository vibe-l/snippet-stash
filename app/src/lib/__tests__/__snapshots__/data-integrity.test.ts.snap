// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Data file integrity > should generate consistent prompts_IDs.json and CSV files 1`] = `"3f48b29b1434c51ec0b7eee01e69e347296c4d28c90fe26134311b5dd8bba08c"`;

exports[`Data file integrity > should generate consistent prompts_IDs.json and CSV files 2`] = `
[
  "word,count",
  "the,221",
  "please,194",
  "and,163",
  "that,108",
  "for,91",
  "follow,91",
  "provide,80",
  "your,78",
  "answer,67",
]
`;

exports[`Data file integrity > should maintain consistent prompts.json structure 1`] = `"898ca25150aa069a50aad30ffdb68a3e8561f16b9b6927c27f3b3f61c15cb5dc"`;

exports[`Data structure validation > should maintain CSV header structure 1`] = `
{
  "hasValidRows": true,
  "header": [
    "word",
    "count",
  ],
  "sampleRowCount": 5,
  "totalRows": 2129,
}
`;

exports[`Data structure validation > should maintain prompts.json schema 1`] = `
{
  "firstItemType": "string",
  "length": 291,
  "sampleKeys": [
    "0",
    "1",
    "10",
    "11",
    "12",
    "13",
    "14",
    "15",
    "16",
    "17",
    "18",
    "19",
    "2",
    "20",
    "21",
    "22",
    "23",
    "24",
    "25",
    "26",
    "27",
    "28",
    "29",
    "3",
    "30",
    "31",
    "32",
    "33",
    "34",
    "35",
    "36",
    "37",
    "38",
    "39",
    "4",
    "40",
    "41",
    "42",
    "43",
    "44",
    "45",
    "46",
    "47",
    "48",
    "49",
    "5",
    "50",
    "51",
    "52",
    "53",
    "54",
    "55",
    "56",
    "57",
    "58",
    "59",
    "6",
    "60",
    "61",
    "62",
    "63",
    "64",
    "65",
    "66",
    "67",
    "68",
    "69",
    "7",
    "70",
    "71",
    "72",
    "73",
    "8",
    "9",
  ],
  "type": "array",
}
`;

exports[`Data structure validation > should maintain prompts_IDs.json schema 1`] = `
{
  "firstItemType": "object",
  "length": 291,
  "sampleKeys": [
    "document",
    "id",
  ],
  "type": "array",
}
`;

exports[`Data structure validation > should track data file counts and sizes 1`] = `
[
  {
    "filename": "prompts.json",
    "sizeBytes": 113596,
    "sizeCategory": "medium",
  },
  {
    "filename": "prompts_IDs.json",
    "sizeBytes": 133894,
    "sizeCategory": "medium",
  },
  {
    "filename": "prompts_doc_word_count.csv",
    "sizeBytes": 22057,
    "sizeCategory": "medium",
  },
]
`;
